diff --git a/utilities/Android.mk b/utilities/Android.mk
index 3d46331..f4ffd90 100755
--- a/utilities/Android.mk
+++ b/utilities/Android.mk
@@ -47,5 +47,5 @@ LOCAL_MODULE_TAGS := optional
 LOCAL_MODULE := libminizip
 LOCAL_CFLAGS := -Dmain=minizip_main -D__ANDROID__ -DIOAPI_NO_64
 LOCAL_C_INCLUDES := external/zlib
-LOCAL_SRC_FILES := ../../../external/zlib/contrib/minizip/minizip.c ../../../external/zlib/contrib/minizip/zip.c ../../../external/zlib/contrib/minizip/ioapi.c
+LOCAL_SRC_FILES := ../../../external/zlib/src/contrib/minizip/minizip.c ../../../external/zlib/src/contrib/minizip/zip.c ../../../external/zlib/src/contrib/minizip/ioapi.c
 include $(BUILD_STATIC_LIBRARY)
