diff --git a/lib/ARCMigrate/Android.mk b/lib/ARCMigrate/Android.mk
index c79b963..7e2555b 100644
--- a/lib/ARCMigrate/Android.mk
+++ b/lib/ARCMigrate/Android.mk
@@ -38,6 +38,7 @@ clang_arc_migrate_SRC_FILES := \
   TransZeroOutPropsInDealloc.cpp
 
 LOCAL_SRC_FILES := $(clang_arc_migrate_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE := libclangARCMigrate
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/AST/Android.mk b/lib/AST/Android.mk
index 7dd021d..39afcba 100644
--- a/lib/AST/Android.mk
+++ b/lib/AST/Android.mk
@@ -79,6 +79,7 @@ clang_ast_SRC_FILES :=	\
 	VTableBuilder.cpp
 
 LOCAL_SRC_FILES := $(clang_ast_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangAST
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Analysis/Android.mk b/lib/Analysis/Android.mk
index fdf7a4e..d5eec94 100644
--- a/lib/Analysis/Android.mk
+++ b/lib/Analysis/Android.mk
@@ -35,6 +35,7 @@ clang_analysis_SRC_FILES := \
 
 
 LOCAL_SRC_FILES := $(clang_analysis_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangAnalysis
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Basic/Android.mk b/lib/Basic/Android.mk
index 8e50d9c..d880fba 100644
--- a/lib/Basic/Android.mk
+++ b/lib/Basic/Android.mk
@@ -41,6 +41,7 @@ clang_basic_SRC_FILES := \
   VersionTuple.cpp
 
 LOCAL_SRC_FILES := $(clang_basic_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangBasic
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/CodeGen/Android.mk b/lib/CodeGen/Android.mk
index 57cba76..afbf9f4 100644
--- a/lib/CodeGen/Android.mk
+++ b/lib/CodeGen/Android.mk
@@ -61,6 +61,7 @@ LOCAL_MODULE_TAGS := optional
 
 LOCAL_SRC_FILES := $(clang_codegen_SRC_FILES)
 TBLGEN_TABLES := $(clang_codegen_TBLGEN_TABLES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 include $(CLANG_HOST_BUILD_MK)
 include $(CLANG_VERSION_INC_MK)
diff --git a/lib/Driver/Android.mk b/lib/Driver/Android.mk
index 7a3ec43..fe618e2 100644
--- a/lib/Driver/Android.mk
+++ b/lib/Driver/Android.mk
@@ -32,6 +32,7 @@ clang_driver_SRC_FILES := \
   WindowsToolChain.cpp
 
 LOCAL_SRC_FILES := $(clang_driver_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE := libclangDriver
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Edit/Android.mk b/lib/Edit/Android.mk
index 765fdb5..e8f64d7 100644
--- a/lib/Edit/Android.mk
+++ b/lib/Edit/Android.mk
@@ -20,6 +20,7 @@ TBLGEN_TABLES := \
   StmtNodes.inc
 
 LOCAL_SRC_FILES := $(clang_edit_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangEdit
 
diff --git a/lib/Frontend/Android.mk b/lib/Frontend/Android.mk
index c5ce90e..47d5fef 100644
--- a/lib/Frontend/Android.mk
+++ b/lib/Frontend/Android.mk
@@ -52,6 +52,7 @@ clang_frontend_SRC_FILES := \
   VerifyDiagnosticConsumer.cpp
 
 LOCAL_SRC_FILES := $(clang_frontend_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangFrontend
 LOCAL_MODULE_TAGS:= optional
diff --git a/lib/FrontendTool/Android.mk b/lib/FrontendTool/Android.mk
index 2496f6b..cd57e6a 100644
--- a/lib/FrontendTool/Android.mk
+++ b/lib/FrontendTool/Android.mk
@@ -18,6 +18,7 @@ clang_frontend_tool_SRC_FILES := \
   ExecuteCompilerInvocation.cpp
 
 LOCAL_SRC_FILES := $(clang_frontend_tool_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 include $(CLANG_HOST_BUILD_MK)
 include $(CLANG_TBLGEN_RULES_MK)
diff --git a/lib/Lex/Android.mk b/lib/Lex/Android.mk
index 96909d2..0faadbb 100644
--- a/lib/Lex/Android.mk
+++ b/lib/Lex/Android.mk
@@ -34,6 +34,7 @@ clang_lex_SRC_FILES := \
   TokenLexer.cpp
 
 LOCAL_SRC_FILES := $(clang_lex_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangLex
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Parse/Android.mk b/lib/Parse/Android.mk
index e8fa743..c309461 100644
--- a/lib/Parse/Android.mk
+++ b/lib/Parse/Android.mk
@@ -32,6 +32,7 @@ clang_parse_SRC_FILES :=  \
 	Parser.cpp
 
 LOCAL_SRC_FILES := $(clang_parse_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangParse
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Rewrite/Frontend/Android.mk b/lib/Rewrite/Frontend/Android.mk
index 8d15425..70c8c26 100644
--- a/lib/Rewrite/Frontend/Android.mk
+++ b/lib/Rewrite/Frontend/Android.mk
@@ -31,6 +31,7 @@ clang_rewrite_frontend_SRC_FILES := \
 
 LOCAL_SRC_FILES := $(clang_rewrite_frontend_SRC_FILES)
 
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 include $(CLANG_HOST_BUILD_MK)
 include $(CLANG_TBLGEN_RULES_MK)
diff --git a/lib/Sema/Android.mk b/lib/Sema/Android.mk
index c7f0e00..74fa3b4 100644
--- a/lib/Sema/Android.mk
+++ b/lib/Sema/Android.mk
@@ -66,6 +66,7 @@ clang_sema_SRC_FILES :=	\
 	TargetAttributesSema.cpp
 
 LOCAL_SRC_FILES := $(clang_sema_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangSema
 LOCAL_MODULE_TAGS := optional
diff --git a/lib/Serialization/Android.mk b/lib/Serialization/Android.mk
index 01656d9..ee2d57b 100644
--- a/lib/Serialization/Android.mk
+++ b/lib/Serialization/Android.mk
@@ -36,7 +36,7 @@ clang_serialization_SRC_FILES :=\
   ModuleManager.cpp
 
 LOCAL_SRC_FILES := $(clang_serialization_SRC_FILES)
-
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 include $(CLANG_HOST_BUILD_MK)
 include $(CLANG_TBLGEN_RULES_MK)
diff --git a/lib/StaticAnalyzer/Checkers/Android.mk b/lib/StaticAnalyzer/Checkers/Android.mk
index b2ad78e..3fa1033 100644
--- a/lib/StaticAnalyzer/Checkers/Android.mk
+++ b/lib/StaticAnalyzer/Checkers/Android.mk
@@ -84,6 +84,7 @@ include $(CLEAR_TBLGEN_VARS)
 TBLGEN_TABLES := $(clang_static_analyzer_checkers_TBLGEN_TABLES)
 
 LOCAL_SRC_FILES := $(clang_static_analyzer_checkers_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangStaticAnalyzerCheckers
 
diff --git a/lib/StaticAnalyzer/Core/Android.mk b/lib/StaticAnalyzer/Core/Android.mk
index 57bfb0b..12a6ceb 100644
--- a/lib/StaticAnalyzer/Core/Android.mk
+++ b/lib/StaticAnalyzer/Core/Android.mk
@@ -56,6 +56,7 @@ include $(CLEAR_TBLGEN_VARS)
 TBLGEN_TABLES := $(clang_static_analyzer_core_TBLGEN_TABLES)
 
 LOCAL_SRC_FILES := $(clang_static_analyzer_core_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_MODULE:= libclangStaticAnalyzerCore
 
diff --git a/lib/StaticAnalyzer/Frontend/Android.mk b/lib/StaticAnalyzer/Frontend/Android.mk
index e7c6664..8f94bf0 100644
--- a/lib/StaticAnalyzer/Frontend/Android.mk
+++ b/lib/StaticAnalyzer/Frontend/Android.mk
@@ -25,6 +25,7 @@ include $(CLEAR_TBLGEN_VARS)
 TBLGEN_TABLES := $(clang_static_analyzer_frontend_TBLGEN_TABLES)
 
 LOCAL_SRC_FILES := $(clang_static_analyzer_frontend_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_C_INCLUDES := $(clang_static_analyzer_frontend_C_INCLUDES)
 
diff --git a/tools/driver/Android.mk b/tools/driver/Android.mk
index 181310d..1938d97 100644
--- a/tools/driver/Android.mk
+++ b/tools/driver/Android.mk
@@ -22,6 +22,7 @@ clang_SRC_FILES := \
   driver.cpp
 
 LOCAL_SRC_FILES := $(clang_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 LOCAL_STATIC_LIBRARIES := \
   libclangFrontendTool \
diff --git a/utils/TableGen/Android.mk b/utils/TableGen/Android.mk
index 3d2ce0e..332fe71 100644
--- a/utils/TableGen/Android.mk
+++ b/utils/TableGen/Android.mk
@@ -16,6 +16,7 @@ include $(CLEAR_VARS)
 LOCAL_MODULE := clang-tblgen
 LOCAL_MODULE_TAGS := optional
 LOCAL_SRC_FILES := $(clang_tablegen_SRC_FILES)
+LOCAL_CFLAGS := -fno-strict-aliasing
 
 REQUIRES_EH := 1
 REQUIRES_RTTI := 1
